examples/caa/caa.ipynb

```diff
@@ -0,0 +1,590 @@
+    "    layers = [15], \n",
+    "    # NOTE: The second last token corresponds to the A/B position\n",
+    "    # which is where we believe the model makes its decision \n",
+    "    read_token_index=-2,RESULTS[f\"steer_0\"] = result\n",
+    "    show_progress=True,\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "SteeringVector(layer_activations={15: tensor([ 0.0644, -0.0403,  0.0907,  ..., -0.1613, -0.0146,  0.0941],\n",
+      "       dtype=torch.float16)}, layer_type='decoder_block')\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(steering_vector)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's sanity check our vector by evaluating the cosine similarity with the ground truth sycophancy vector."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Cosine similarity: 0.995\n"
+     ]
+    }
+   ],
+   "source": [
+    "original_steering_vector = torch.load(f'vec_layer_15_Llama-2-{model_size}-chat-hf.pt')\n",
+    "our_steering_vector = steering_vector.layer_activations[15]\n",
+    "\n",
+    "def calculate_cosine_similarity(a: torch.Tensor, b: torch.Tensor) -> float:\n",
+    "    \"\"\" Calculate cosine similarity between two vectors \"\"\"\n",
+    "    return torch.sum(a * b) / (a.norm() * b.norm())\n",
+    "\n",
+    "print(f\"Cosine similarity: {calculate_cosine_similarity(original_steering_vector, our_steering_vector):.3f}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Steer with Steering Vectors"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To apply steering vectors to the model, we must have some way of modifying the model's forward pass. The approach taken in the official codebases for Representation Engineering and Contrastive Activation Addition both do this by wrapping the model's blocks. This approach is conceptually simpler but has notable limitations, e.g. being unable to work for other models\n",
+    " \n",
+    "The `steering_vectors` library uses Pytorch hooks instead of model wrappers to modify the underlying model's forward pass. You can read more about hooks at the [official PyTorch documentation](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_hook.html). "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "+1 steered model: 0.716\n",
+      "Unsteered model: 0.685\n"
+     ]
+    }
+   ],
+   "source": [
+    "hook = steering_vector.patch_activations(\n",
```

Comment:
nit: we should just skip the `patch_activations()` version and go straight to the `steering_vector.apply()` approach, that's what people should use in most cases, so no need to show them the more complicated thing in a demo when the simpler approach works better.

Reply:
Yeah I agree we should put the preferred usage first. However, I imagine this is going to be most peoples' first introduction to PyTorch hooks and I want to at least make it clear how we use `hook` internally

Reply:
I ended up writing a pre-amble discussing pytorch hooks with an inline code block for the manual hook example. Hopefully the fact that it's not executable is a strong enough indication that this isn't preferred.

Reply:
I would still argue that we shouldn't even tell them about the `patch_activations()` version at all in this notebook, since that seems like it can only add confusion. The `apply()` version is still using hooks, it's just a nicer interface to work with and can do all the same things. That being said, it's up to you.